src/analysis/fourier/fourier_add_circle.lean

```diff
@@ -0,0 +1,330 @@
+import topology.continuous_function.stone_weierstrass
+
+/-!
+
+# Fourier analysis on the additive circle
+
+This file contains basic results on Fourier series for functions on the additive circle
+`add_circle T = ℝ / ℤ • T`.
+
+## Main definitions
+
+* `haar_add_circle`, Haar measure on `add_circle T`, normalized to have total measure `1`. (Note
+  that this is not the same normalisation as the standard measure defined in `integral.periodic`,
+  so we do not declare it as a `measure_space` instance, to avoid confusion.)
+* for `n : ℤ`, `fourier n` is the monomial `λ x, exp (2 π i n x / T)`, bundled as a continuous map
+  from `add_circle T` to `ℂ`.
+* for `n : ℤ` and `p : ℝ≥0∞`, `fourier_Lp p n` is an abbreviation for the monomial `fourier n`
+  considered as an element of the Lᵖ-space `Lp ℂ p haar_add_circle`, via the embedding
+  `continuous_map.to_Lp`.
+* `fourier_series` is the canonical isometric isomorphism from `Lp ℂ 2 haar_add_circle` to
+  `ℓ²(ℤ, ℂ)` induced by taking Fourier coefficients.
+
+## Main statements
+
+The theorem `span_fourier_closure_eq_top` states that the span of the monomials `fourier n` is
+dense in `C(add_circle T, ℂ)`, i.e. that its `submodule.topological_closure` is `⊤`.  This follows
+from the Stone-Weierstrass theorem after checking that it is a subalgebra, closed under conjugation,
+and separates points.
+
+The theorem `span_fourier_Lp_closure_eq_top` states that for `1 ≤ p < ∞` the span of the monomials
+`fourier_Lp` is dense in the Lᵖ space of `add_circle T`, i.e. that its
+`submodule.topological_closure` is `⊤`. This follows from the previous theorem using general theory
+on approximation of Lᵖ functions by continuous functions.
+
+The theorem `orthonormal_fourier` states that the monomials `fourier_Lp 2 n` form an orthonormal set
+(in the L² space of `add_circle T` with respect to `haar_add_circle`).
+
+The last two results together provide that the functions `fourier_Lp 2 n` form a Hilbert basis for
+L²; this is named as `fourier_series`.
+
+Parseval's identity, `tsum_sq_fourier_series_repr`, is a direct consequence of the construction of
+this Hilbert basis.
+-/
+
+noncomputable theory
+open_locale ennreal complex_conjugate classical real
+open topological_space continuous_map measure_theory measure_theory.measure algebra submodule set
+
+variables {T : ℝ}
+
+namespace add_circle
+
+/-! ### Map from `add_circle` to `circle` -/
+
+lemma scaled_exp_map_periodic :
+  function.periodic (λ x, exp_map_circle (2 * π / T * x)) T :=
+begin
+  -- The case T = 0 is not interesting, but it is true, so we prove it to save hypotheses
+  rcases eq_or_ne T 0 with rfl | hT,
+  { intro x, simp },
+  { intro x, simp_rw mul_add, rw [div_mul_cancel _ hT, periodic_exp_map_circle] }
+end
+
+/-- The canonical map `λ x, exp (2 π i x / T)` from `ℝ / ℤ • T` to the unit circle in `ℂ`.
+If `T = 0` we understand this as the constant function 1. -/
+def to_circle := (@scaled_exp_map_periodic T).lift
```

Comment:
Maybe, but which? There are two `circle` files (`analysis.complex.circle` and `analysis.special_functions.complex.circle`) and I don't entirely understand the reasons for the split. The latter of those already has some results about `exp_map_circle` applied to a `real.angle`, which is a special case of `add_circle`, so maybe that is the right place for these too.

(One could clearly add many more lemmas about `add_circle.to_circle`, but I made a concerted effort here to only introduce the results that were absolutely necessary for Fourier theory rather than getting side-tracked.)

Reply:
Looking into this a little more, there is indeed quite a maze of related constructions.  I think that in some future reorganization we should write out some theory of the map `(λ x : ℝ, exp (x * I))` from `ℝ` to `ℂ`: things that are currently proved for `exp_map_circle`, like
```lean
lemma complex.exp_mul_I_eq_exp_mul_I {x y : ℝ} :
  exp (x * I) = exp (y * I) ↔ ∃ m : ℤ, x = y + m * (2 * π) :=
begin
  rw [exp_eq_exp_iff_exists_int],
  refine exists_congr (λ n, _),
  rw [← mul_assoc, ← add_mul, mul_left_inj' I_ne_zero, ← of_real_one, ← of_real_bit0,
    ← of_real_mul, ← of_real_int_cast, ← of_real_mul, ← of_real_add, of_real_inj]
end

lemma complex.periodic_exp_mul_I : function.periodic (λ x : ℝ, exp (x * I)) (2 * π) :=
λ z, complex.exp_mul_I_eq_exp_mul_I.2 ⟨1, by rw [int.cast_one, one_mul]⟩

...
``` 

Then this Fourier file can use that theory directly, with no need to mention `circle` at all, and the `to_circle` map will disappear (in this file).  And the `exp_map_circle` lemmas can also have proofs routed via that theory.

Given that plan, there's no point in optimizing the development here, so let's leave your presentation as-is.